dotnet/Query/Query/Query.cs

```diff
@@ -0,0 +1,550 @@
+    {
+        // Java: findTestCaseBy(testStepStarted).map(TestCase::getPickleId).map(pickleById::get)
+        var testCaseStarted = FindTestCaseStartedBy(testStepStarted);
+        if (testCaseStarted != null)
+        {
+            return FindPickleBy(testCaseStarted);
+        }
+        return null;
+    }
+
+    public TestCase? FindTestCaseBy(TestCaseStarted testCaseStarted)
+    {
+        // Java: testCaseById.get(testCaseStarted.getTestCaseId())
+        if (_testCaseById.TryGetValue(testCaseStarted.TestCaseId, out var testCase))
+        {
+            return testCase;
+        }
+        return null;
+    }
+
+    public TestCaseStarted? FindTestCaseStartedBy(TestStepStarted testStepStarted)
+    {
+        // Java: testCaseStartedById.get(testStepStarted.getTestCaseStartedId())
+        return _testCaseStartedById.TryGetValue(testStepStarted.TestCaseStartedId, out var tcs) ? tcs : null;
+    }
+
+    public TestCase? FindTestCaseBy(TestStepStarted testStepStarted)
+    {
+        // Java: findTestCaseStartedBy(testStepStarted).flatMap(this::findTestCaseBy)
+        var testCaseStarted = FindTestCaseStartedBy(testStepStarted);
+        return testCaseStarted != null ? FindTestCaseBy(testCaseStarted) : null;
+    }
+
+    public TestStep? FindTestStepBy(TestStepStarted testStepStarted)
+    {
+        // Java: testStepById.get(testStepStarted.getTestStepId())
+        return _testStepById.TryGetValue(testStepStarted.TestStepId, out var testStep) ? testStep : null;
+    }
+
+    public TestStep? FindTestStepBy(TestStepFinished testStepFinished)
+    {
+        // Java: testStepById.get(testStepFinished.getTestStepId())
+        return _testStepById.TryGetValue(testStepFinished.TestStepId, out var testStep) ? testStep : null;
+    }
+
+    public PickleStep? FindPickleStepBy(TestStep testStep)
+    {
+        // Java: testStep.getPickleStepId().map(pickleStepById::get)
+        if (!string.IsNullOrEmpty(testStep.PickleStepId))
+        {
+            if (_pickleStepById.TryGetValue(testStep.PickleStepId, out var pickleStep))
+            {
+                return pickleStep;
+            }
+        }
+        return null;
+    }
+
+    public Step? FindStepBy(PickleStep pickleStep)
+    {
+        // Java: String stepId = pickleStep.getAstNodeIds().get(0); stepById.get(stepId)
+        if (pickleStep.AstNodeIds != null && pickleStep.AstNodeIds.Count > 0)
+        {
+            var stepId = pickleStep.AstNodeIds[0];
+            if (!string.IsNullOrEmpty(stepId) && _stepById.TryGetValue(stepId, out var step))
+            {
+                return step;
+            }
+        }
+        return null;
+    }
+
+    public System.TimeSpan? FindTestCaseDurationBy(TestCaseStarted testCaseStarted)
+    {
+        var started = testCaseStarted.Timestamp;
+        var finished = FindTestCaseFinishedBy(testCaseStarted)?.Timestamp;
+        if (finished != null)
+        {
+            var startTime = started.ToDateTimeOffset();
```

Comment:
So I'm not 100% on .Net, but Messages has a [`ToDateTime`](https://github.com/cucumber/messages/blob/47defbf650bec0b3e7525dd47698f31784d39b9f/dotnet/Cucumber.Messages/Converters.cs#L31) method and [`DateTime` supports substraction into a timespan](https://learn.microsoft.com/en-us/dotnet/api/system.datetime.op_subtraction?view=net-9.0#system-datetime-op-subtraction(system-datetime-system-datetime)).

Is the `ToDateTimeOffset` needed?